File: app/src/main/java/com/ahmedabdelmeged/pagingwithrxjava/java/ui/MainActivity.java
package com.ahmedabdelmeged.pagingwithrxjava.java.ui;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.ahmedabdelmeged.pagingwithrxjava.core.BaseActivity;
import com.ahmedabdelmeged.pagingwithrxjava.databinding.ActivityMainBinding;
import com.ahmedabdelmeged.pagingwithrxjava.java.adapter.RetryCallback;
import com.ahmedabdelmeged.pagingwithrxjava.java.adapter.UserAdapter;
import com.ahmedabdelmeged.pagingwithrxjava.java.data.NetworkState;
import com.ahmedabdelmeged.pagingwithrxjava.java.data.Status;

public class MainActivity extends BaseActivity<ActivityMainBinding> implements RetryCallback {

    private UsersViewModel usersViewModel;

    private UserAdapter userAdapter;

    @NonNull
    @Override
    public ActivityMainBinding getViewBinding() {
        return ActivityMainBinding.inflate(getLayoutInflater());
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        usersViewModel = ViewModelProviders.of(this).get(UsersViewModel.class);
        initAdapter();
        initSwipeToRefresh();
    }

    private void initAdapter() {
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false);
        userAdapter = new UserAdapter(this);
        getBinding().usersRecyclerView.setLayoutManager(linearLayoutManager);
        getBinding().usersRecyclerView.setAdapter(userAdapter);
        usersViewModel.userList.observe(this, userAdapter::submitList);
        usersViewModel.getNetworkState().observe(this, userAdapter::setNetworkState);
    }

    /**
     * Init swipe to refresh and enable pull to refresh only when there are items in the adapter
     */
    private void initSwipeToRefresh() {
        usersViewModel.getRefreshState().observe(this, networkState -> {
            if (networkState != null) {
                if (userAdapter.getCurrentList() != null) {
                    if (userAdapter.getCurrentList().size() > 0) {
                        getBinding().usersSwipeRefreshLayout.setRefreshing(
                                networkState.getStatus() == NetworkState.LOADING.getStatus());
                    } else {
                        setInitialLoadingState(networkState);
                    }
                } else {
                    setInitialLoadingState(networkState);
                }
            }
        });
        getBinding().usersSwipeRefreshLayout.setOnRefreshListener(() -> usersViewModel.refresh());
    }

    /**
     * Show the current network state for the first load when the user list
     * in the adapter is empty and disable swipe to scroll at the first loading
     *
     * @param networkState the new network state
     */
    private void setInitialLoadingState(NetworkState networkState) {
        //error message
        getBinding().containerLoading.errorMessageTextView.setVisibility(networkState.getMessage() != null ? View.VISIBLE : View.GONE);
        if (networkState.getMessage() != null) {
            getBinding().containerLoading.errorMessageTextView.setText(networkState.getMessage());
        }

        //loading and retry
        getBinding().containerLoading.retryLoadingButton.setVisibility(networkState.getStatus() == Status.FAILED ? View.VISIBLE : View.GONE);
        getBinding().containerLoading.loadingProgressBar.setVisibility(networkState.getStatus() == Status.RUNNING ? View.VISIBLE : View.GONE);
        getBinding().containerLoading.retryLoadingButton.setOnClickListener(v -> usersViewModel.retry());

        getBinding().usersSwipeRefreshLayout.setEnabled(networkState.getStatus() == Status.SUCCESS);
    }

    @Override
    public void retry() {
        usersViewModel.retry();
    }

}

File: app/src/main/java/com/ahmedabdelmeged/pagingwithrxjava/kotlin/ui/MainActivity.kt
package com.ahmedabdelmeged.pagingwithrxjava.kotlin.ui

import android.os.Bundle
import android.view.View
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.ahmedabdelmeged.pagingwithrxjava.core.BaseActivity
import com.ahmedabdelmeged.pagingwithrxjava.databinding.ActivityMainBinding
import com.ahmedabdelmeged.pagingwithrxjava.kotlin.adapter.UserAdapter
import com.ahmedabdelmeged.pagingwithrxjava.kotlin.data.NetworkState
import com.ahmedabdelmeged.pagingwithrxjava.kotlin.data.Status

class MainActivity : BaseActivity<ActivityMainBinding>() {

    private lateinit var userAdapter: UserAdapter

    private val usersViewModel: UsersViewModel by lazy {
        ViewModelProviders.of(this)[UsersViewModel::class.java]
    }

    override fun getViewBinding(): ActivityMainBinding {
        return ActivityMainBinding.inflate(layoutInflater)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        initAdapter()
        initSwipeToRefresh()
    }

    private fun initAdapter() {
        val linearLayoutManager = LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false)
        userAdapter = UserAdapter {
            usersViewModel.retry()
        }
        binding.usersRecyclerView.layoutManager = linearLayoutManager
        binding.usersRecyclerView.adapter = userAdapter
        usersViewModel.userList.observe(this) { userAdapter.submitList(it) }
        usersViewModel.getNetworkState().observe(this) { userAdapter.setNetworkState(it) }
    }

    /**
     * Init swipe to refresh and enable pull to refresh only when there are items in the adapter
     */
    private fun initSwipeToRefresh() {
        usersViewModel.getRefreshState().observe(this) { networkState ->
            if (userAdapter.currentList != null) {
                if (userAdapter.currentList!!.size > 0) {
                    binding.usersSwipeRefreshLayout.isRefreshing =
                        networkState?.status == NetworkState.LOADING.status
                } else {
                    setInitialLoadingState(networkState)
                }
            } else {
                setInitialLoadingState(networkState)
            }
        }
        binding.usersSwipeRefreshLayout.setOnRefreshListener { usersViewModel.refresh() }
    }

    /**
     * Show the current network state for the first load when the user list
     * in the adapter is empty and disable swipe to scroll at the first loading
     *
     * @param networkState the new network state
     */
    private fun setInitialLoadingState(networkState: NetworkState?) {
        //error message
        binding.containerLoading.errorMessageTextView.visibility = if (networkState?.message != null) View.VISIBLE else View.GONE
        if (networkState?.message != null) {
            binding.containerLoading.errorMessageTextView.text = networkState.message
        }

        //loading and retry
        binding.containerLoading.retryLoadingButton.visibility = if (networkState?.status == Status.FAILED) View.VISIBLE else View.GONE
        binding.containerLoading.loadingProgressBar.visibility = if (networkState?.status == Status.RUNNING) View.VISIBLE else View.GONE

        binding.usersSwipeRefreshLayout.isEnabled = networkState?.status == Status.SUCCESS
        binding.containerLoading.retryLoadingButton.setOnClickListener { usersViewModel.retry() }
    }

}
